crates/bevy_reflect/src/type_registry.rs

```diff
@@ -350,11 +350,103 @@ impl<T: for<'a> Deserialize<'a> + Reflect> FromType<T> for ReflectDeserialize {
     }
 }
 
+#[derive(Clone)]
+pub struct ReflectFromPtr {
+    type_id: TypeId,
+    to_reflect: for<'a> unsafe fn(*const (), lt: PhantomData<&'a ()>) -> &'a dyn Reflect,
+    to_reflect_mut: for<'a> unsafe fn(*mut (), lt: PhantomData<&'a ()>) -> &'a mut dyn Reflect,
+}
+
+impl ReflectFromPtr {
+    /// # Safety:
+    /// - `val` must be a pointer to a value of the type that the [`ReflectFromPtr`] was constructed for
+    /// - the lifetime `'a` of the return type can be arbitrarily chosen by the caller, who must ensure that during
+    /// that lifetime, `val` is valid
+    pub unsafe fn to_reflect_ptr<'a>(&self, val: *const ()) -> &'a dyn Reflect {
```

Comment:
Slight nitpick: I wonder if these methods should be `as_***` rather than `to_***` to help signify that they're "casting" rather than taking ownership and converting (unless my assumption about that is wrong).

Or would this cause confusion with the similarly named methods in #4350?

Reply:
I agree, looking back at the names `to_reflect_ptr` doesn't really make a lot of sense, especially since it returns a reference.

I also realized that the method could be `to_reflect_ptr(&self, val: *const()) -> *const dyn Reflect` which gets rid of the footgun where the caller can choose an arbitrary lifetime.

I changed the names to this:
```rs
pub unsafe fn as_reflect_ptr(&self, val: *const ()) -> *const dyn Reflect;
pub unsafe fn as_reflect_ptr_mut(&self, val: *const ()) -> *mut dyn Reflect;
pub fn as_reflect<'a, T: Reflect>(&self, val: &'a T) -> &'a dyn Reflect;
pub fn as_reflect_mut<'a, T: Reflect>(&self, val: &'a mut T) -> &'a mut dyn Reflect;
```